packages/mangrove.js/src/tracker/blockManager.ts

```diff
@@ -0,0 +1,566 @@
+      if (res.error) {
+        /* initialize call failed retry later by adding it back to the set */
+        this.waitingToBeInitializedSet.add(address);
+      } else {
+        logger.debug(
+          `subscriberInitialize() ${address} ${getStringBlock(res.block)}`
+        );
+      }
+    }
+
+    return toInitialize;
+  }
+
+  /**
+   * For each logs find if there is a matching subscriber, then call handle log on the subscriber
+   */
+  private applyLogs(logs: Log[]) {
+    if (this.subscibedAddresses.length === 0) {
+      return;
+    }
+
+    for (const log of logs) {
+      const checksumAddress = getAddress(log.address);
+      log.address = checksumAddress; // DIRTY: Maybe do it at the RPC level ?
+
+      const subscriber = this.subscribersByAddress[checksumAddress];
+      subscriber.handleLog(log);
+      logger.debug(
+        `handleLog() ${log.address} (${log.blockHash}, ${log.blockNumber})`
+      );
+    }
+  }
+
+  /**
+   * Call rollback subscriber on all subscriber with lastSeenEventBlockNumber > block.number,
+   * schedule re-initialize for subscriber with initializedAt > block.number
+   */
+  private rollbackSubscribers(block: BlockManager.Block) {
+    for (const [address, subscriber] of Object.entries(
+      this.subscribersByAddress
+    )) {
+      if (subscriber.initializedAt.number > block.number) {
+        /* subscriber has been initialized at a block newer than block
+         * it needs to be initialized again.
+         **/
+        this.waitingToBeInitializedSet.add(address);
+        logger.debug(
+          `addToInitializeList() ${address} ${getStringBlock(
+            subscriber.initializedAt
+          )} ${getStringBlock(block)}`
+        );
+      } else if (subscriber.lastSeenEventBlockNumber > block.number) {
+        subscriber.rollback(block);
+        logger.debug(`rollback() ${address} ${getStringBlock(block)}`);
+      }
+    }
+  }
+
+  /**
+   * Verify that subscriber has been initialized with a block that we know in cache
+   */
+  private verifySubscribers(initializedSubscribers: string[]) {
+    for (const address of initializedSubscribers) {
+      const subscriber = this.subscribersByAddress[address];
+      if (subscriber.initializedAt) {
+        const cachedBlock =
+          this.blocksByNumber[subscriber.initializedAt.number];
+
+        if (cachedBlock.hash !== subscriber.initializedAt.hash) {
+          /* subscriber initializedAt block is different from block in cache
+           *  a reorg happened during initialiation. Retry initialiation later.
+           **/
+          this.waitingToBeInitializedSet.add(address);
+          logger.debug(
+            `verifySubscriber() detected wrong subscriber ${address} ${getStringBlock(
+              subscriber.initializedAt
+            )} !== ${getStringBlock(cachedBlock)}`
+          );
+        }
+      }
+    }
+  }
+
+  /**
+   * Add new block in BlockManager cache detect reorganization and ensure that cache is consistent
```

Comment:
I think some breaks/commas would be good here, eg:
```suggestion
   * Add new block in BlockManager cache, detect reorganization, and ensure that cache is consistent
```